W-12560426-k8-connected-vs-local-duke (#175)
mulesoft/docs-gateway: gateway/1.4/modules/ROOT/pages/_partials/task-reg-run-flex-gateway.adoc

// tag::reg-command-intro[]
To register a Flex Gateway with Anypoint Platform, you must enter the registration command and then the start command. Each command includes information specific to your Anypoint Platform instance and *must be updated before* executing. See <<Before You Begin>> for more information on how to find the information you will need.
//end::reg-command-intro[]
//
// tag::k8s-connected-intro[]

> In this mode, the gateway is managed entirely through Anypoint Platform, where you can: 

Reviewer: I suggest using a softer term than `entirely` because there are some configurations that must be provided as local files/k8s resources.
 
```suggestion
In this mode, the gateway is managed mainly through Anypoint Platform, where you can: 
```

Author: @seblaz If we use "mainly" here (which we certainly can), we'll need to very specific about what is managed where. Words like "mainly" makes the readers ask questions.

* Deploy APIs to your cluster.
* Apply policies and other configurations.

> * Create all resources through Anypoint Platform. 

Reviewer: There are some configuration that must be provided as local files/k8s resources (for example, the `Configuration` resource).
```suggestion
```

Author: @seblaz We can delete that inaccurate bullet, but readers will almost certainly want to know what types of resources they can create (and can't created) via Anypoint Platform.

Is there something we can say here about what they can create through Anypoint Platform in connected mode (compared to what they can't create through Anypoint Platform when in local mode).

> Resource creation through `kubectl apply` is restricted to specific resource types.

Author: which "resource types"? 

I think we're contrasting this with the use of that command for local mode, so important to clarify.

Reviewer: The following resources are allowed to be created in connected mode:
- Extension
- Service
- PolicyBinding of types `tls`, `tls-inbound` and `tls-outbound`

Reviewer: The following resources are allowed to be created in connected mode:
- Configuration
- Secret
- Extension
- Service
- PolicyBinding of types `tls`, `tls-inbound` and `tls-outbound`

Author: @seblaz Can you create any of these resources listed here via Anypoint Platform?

Reviewer: It's a bit complicated:
1. `Configuration` can't be created via Anypoint
2. `Secret` is only used for registration, so I think it's not important. 
3. `Extension` can be created via Anypoint with custom policies.
4. `Service` is created via Anypoint when deploying apis, but if a custom policy wants to make an http request it needs to create another `Service` which can't be created via Anypoint.
5. `PolicyBinding`s can be created via Anypoint

Author: thank you!

> Resource creation through `kubectl apply` is restricted to specific resource types.

Reviewer: In targets that are not k8s (like docker and vms) `kubectl apply` is not used

Author: should we mention ☝️ in the docs?

Reviewer: I think we should try to say something that includes all targets. Maybe something like `Resource creation is restricted to specific resource types.` omitting how they would be created, because that is different in each target

Reviewer: I think that later in the docs we specify how this works in each target, so we might not want to go into those details here

> Resource creation through `kubectl apply` is restricted to specific resource types.

Author: ```suggestion
Resource creation through `kubectl apply` is restricted to specific resource types:

* Configuration
* Secret
* Extension
* Service
* PolicyBinding of types `tls`, `tls-inbound`, and `tls-outbound`
```

Reviewer: LGTM

// end::k8s-connected-intro[]
//
// tag::k8s-local-intro[]

> In Local Mode, Flex Gateway acts as an ingress controller that manages external access to your Kubernetes cluster. When using this mode, you must know what resources to create and apply, and use `kubectl apply` to deploy resources such as APIs, policies, or Flex Gateway-related resources to your gateway. Local Mode also supports the Ingress resources from Kubernetes, which are configured through a YAML file with properties, such as `apiVersion`, `kind`, `metadata`, `spec` for configuring a load balancer or proxy server, and `rules` for directing HTTP and HTTPS traffic (see https://kubernetes.io/docs/concepts/services-networking/ingress/[Kubernetes documentation^] on Ingress resources).

Author: For local mode, are we applying `kubectl apply` for ALL resources types, unlike in connected mode, where it's applied only for "restricted resource types"? The contrast isn't entirely clear to me in the G-doc (attached to W-12560426).

Reviewer: I can give you more context about this. For example, in connected mode the api instances are configured via Anypoint, therefore we don't allow to create api instances through local resources in that mode. In the other hand, in local mode the only option is to create the api instances through local resources, therefore that is allowed.

The conclusion is that in local mode since the only source of configuration are local resources, then all resources are allowed to be created through that method. In connected mode some are not allowed because they can be created via Anypoint

Author: thank you!

> In Local Mode, Flex Gateway acts as an ingress controller that manages external access to your Kubernetes cluster. When using this mode, you must know what resources to create and apply, and use `kubectl apply` to deploy resources such as APIs, policies, or Flex Gateway-related resources to your gateway. Local Mode also supports the Ingress resources from Kubernetes, which are configured through a YAML file with properties, such as `apiVersion`, `kind`, `metadata`, `spec` for configuring a load balancer or proxy server, and `rules` for directing HTTP and HTTPS traffic (see https://kubernetes.io/docs/concepts/services-networking/ingress/[Kubernetes documentation^] on Ingress resources).

Reviewer: I suggest using the ingress controller use case as an example, but not say that it's the only one because there are other use cases covered by flex in local mode
```suggestion
In Local Mode, Flex Gateway can act as an ingress controller that manages external access to your Kubernetes cluster. When using this mode, you must know what resources to create and apply, and use `kubectl apply` to deploy resources such as APIs, policies, or Flex Gateway-related resources to your gateway. Local Mode also supports the Ingress resources from Kubernetes, which are configured through a YAML file with properties, such as `apiVersion`, `kind`, `metadata`, `spec` for configuring a load balancer or proxy server, and `rules` for directing HTTP and HTTPS traffic (see https://kubernetes.io/docs/concepts/services-networking/ingress/[Kubernetes documentation^] on Ingress resources).
```

Author: @seblaz what other ways can Flex act besides an ingress controller? Is ingress controller something we recommend in terms of Flex specifically? Do we care how Flex acts? What ways do we recommend, if any?

Author: Basically, I would like to be able to say something that provides a bit of guidance about the use case we thing customers are likely to have for Flex, like "Flex Gateway typically acts as an ingress controller" or, for example (just an example I am making up), that we recommend using it as an ingress controller when in local mode, or whatever guidance we can give.

Author: @seblaz What's the common use case, and do we/MuleSoft have any opinion about how people should use Flex in local mode? And do people usually use it as an ingress controller in this case vs. in connected mode. And how do they usually use it in connected mode? Thx!

Reviewer: > what other ways can Flex act besides an ingress controller? 

Mainly, as an ingress controller or as a standalone gateway

> Is ingress controller something we recommend in terms of Flex specifically?

It's probably the most common use case for local mode, so I think it makes sense to mention it here

> What ways do we recommend, if any?

I don't think we can make a recommendation about the use cases, it depends on the context where the users want to use it

> Basically, I would like to be able to say something that provides a bit of guidance about the use case we thing customers are likely to have for Flex, like "Flex Gateway typically acts as an ingress controller" or, for example (just an example I am making up), that we recommend using it as an ingress controller when in local mode, or whatever guidance we can give.

"Flex Gateway can act as an ingress controller or standalone gateway" covers most cases, so I think that should be good enough

> And how do they usually use it in connected mode? 

Usually as an ingress (note that this is not the same as ingress controller), or as a standalone gateway as well

> In Local Mode, Flex Gateway acts as an ingress controller that manages external access to your Kubernetes cluster. When using this mode, you must know what resources to create and apply, and use `kubectl apply` to deploy resources such as APIs, policies, or Flex Gateway-related resources to your gateway. Local Mode also supports the Ingress resources from Kubernetes, which are configured through a YAML file with properties, such as `apiVersion`, `kind`, `metadata`, `spec` for configuring a load balancer or proxy server, and `rules` for directing HTTP and HTTPS traffic (see https://kubernetes.io/docs/concepts/services-networking/ingress/[Kubernetes documentation^] on Ingress resources).

Reviewer: In targets that are not k8s (like docker and vms) `kubectl apply` is not used

// tag::reg-command-intro[]
To register a Flex Gateway with Anypoint Platform, you must enter the registration command and then the start command. Each command includes information specific to your Anypoint Platform instance and *must be updated before* executing. See <<Before You Begin>> for more information on how to find the information you will need.
//end::reg-command-intro[]
//
// tag::k8s-connected-intro[]
In this mode, the gateway is managed mostly through Anypoint Platform, where you can: 

* Deploy APIs to your cluster.
* Apply policies and other configurations.
* Create resources through Anypoint Platform _with the exception_ of Configuration and Service resources. 

In addition, resource creation through `kubectl apply` is restricted to the following resource types:

* Configuration
* Secret
* Extension
* Service
* PolicyBinding of types `tls`, `tls-inbound`, and `tls-outbound`

// end::k8s-connected-intro[]
//
// tag::k8s-local-intro[]

> In Local Mode, Flex Gateway acts as an ingress controller that manages external access to your Kubernetes cluster. When using this mode, you must know what resources to create and apply, and use `kubectl apply` to deploy resources such as APIs, policies, or Flex Gateway-related resources to your gateway. Local Mode also supports the Ingress resources from Kubernetes, which are configured through a YAML file with properties, such as `apiVersion`, `kind`, `metadata`, `spec` for configuring a load balancer or proxy server, and `rules` for directing HTTP and HTTPS traffic (see https://kubernetes.io/docs/concepts/services-networking/ingress/[Kubernetes documentation^] on Ingress resources).

Author: ```suggestion
In Local Mode, Flex Gateway typically acts as an ingress controller that manages external access to your Kubernetes cluster. When using this mode, you must know what resources to create and apply, and use `kubectl apply` to deploy Kubernetes targets and resources such as APIs, policies, or Flex Gateway-related resources to your gateway. 

Local Mode also supports the Ingress resources from Kubernetes, which are configured through a YAML file with properties, such as `apiVersion`, `kind`, `metadata`, `spec` for configuring a load balancer or proxy server, and `rules` for directing HTTP and HTTPS traffic (see https://kubernetes.io/docs/concepts/services-networking/ingress/[Kubernetes documentation^] on Ingress resources).
```

// tag::reg-command-intro[]
To register a Flex Gateway with Anypoint Platform, you must enter the registration command and then the start command. Each command includes information specific to your Anypoint Platform instance and *must be updated before* executing. See <<Before You Begin>> for more information on how to find the information you will need.
//end::reg-command-intro[]
//
// tag::k8s-connected-intro[]
In this mode, the gateway is managed mostly through Anypoint Platform, where you can: 

* Deploy APIs to your cluster.
* Apply policies and other configurations.

> * Create resources through Anypoint Platform _with the exception_ of Configuration and Service resources. 

Author: @seblaz What's are examples of a Service  and Configuration resources?

Reviewer: We have examples in [these docs](https://docs.mulesoft.com/gateway/latest/flex-local-configuration-reference-guide). Although they are under "Manage Flex Gateway in Local Mode" they work for connected mode as well

Author: Hmm, I wonder if we need to document that in both places?

Author: I added a link to that page below. I'm not sure what to do about the fact that it's in the Local Mode config section of the docs...

In addition, resource creation through `kubectl apply` is restricted to the following resource types:

* Configuration
* Secret
* Extension
* Service

> * PolicyBinding of types `tls`, `tls-inbound`, and `tls-outbound`

Reviewer: IMHO, if this is an introduction we can remove this list because we are specifying the resources that can't be created via Anypoint in the previous paragraph

Author: Well, I'm not sure because the previous paragraph qualifies a condition where you can't use Anypoint Platform, but the second here is about where you CAN (but in some cases don't have to) use kubectl apply. So I think we're talking about different things where there is some overlap. Am I misunderstanding something?

Reviewer: Yes you got it right, we can leave the list then

> 

Author: ```suggestion

For more info about resources, see xref:flex-local-configuration-reference-guide.adoc[].
```

// end::k8s-connected-intro[]
//
// tag::k8s-local-intro[]

> In Local Mode, Flex Gateway typically acts as an ingress controller that manages external access to your Kubernetes cluster. When using this mode, you must know what resources to create and apply, and use `kubectl apply` to deploy Kubernetes targets and resources such as APIs, policies, or Flex Gateway-related resources to your gateway. 

Author: this one qualifies that we're talking about using the command for _Kubernetes_ targets and resources

// tag::reg-command-intro[]
To register a Flex Gateway with Anypoint Platform, you must enter the registration command and then the start command. Each command includes information specific to your Anypoint Platform instance and *must be updated before* executing. See <<Before You Begin>> for more information on how to find the information you will need.
//end::reg-command-intro[]
//
// tag::k8s-connected-intro[]
In Connected Mode, Flex Gateway typically serves as an ingress or a standalone gateway that is managed mostly through Anypoint Platform, where you can: 

* Deploy APIs to your cluster.
* Apply policies and other configurations.

> * Create resources through Anypoint Platform _with the exception_ of Configuration and Service resources. 

Author: ```suggestion
* Create resources (_with the exception_ of Configuration and Service resources). 
```
